lib/llvmopencl/Workgroup.cc

```diff
@@ -195,6 +195,53 @@ class WorkgroupImpl {
   unsigned long DeviceMaxWItemSizes[3];
 };
 
+/// Convert address space casts through integer casts to proper address space
+/// cast instructions.
+///
+/// At least the HIP/CUDA frontend since LLVM 19 seems to produce AS casts
+/// between generic and global ASs through ptrtoint - inttoptr conversion chains
+/// when targeting SPIR-V. It confuses the alias analyzer and leads to lack of
+/// utilization of "restrict" pointer information. It should be fixed upstream,
+/// but while that happens, let's just peephole convert it.
+static bool cleanupAddressSpaceCasts(Function &F) {
+
+  bool Changed = false;
+  llvm::IRBuilder<> Builder(F.getContext());
+  for (auto &BB : F) {
+    for (BasicBlock::iterator BI = BB.begin(), BE = BB.end(); BI != BE;) {
+
+      // The casts look like this:
+      // %55 = ptrtoint ptr addrspace(1) %7 to i64
+      // %56 = inttoptr i64 %55 to ptr addrspace(4)
+
+      PtrToIntInst *PtrToInt = dyn_cast<PtrToIntInst>(&*BI);
+      if (PtrToInt != nullptr &&
+          (PtrToInt->getPointerAddressSpace() == SPIR_ADDRESS_SPACE_GLOBAL ||
+           PtrToInt->getPointerAddressSpace() == SPIR_ADDRESS_SPACE_GENERIC) &&
+          PtrToInt->hasOneUse()) {
```

Comment:
This pass potentially misses the following cast pattern:
```
%1 = ptrtoint ptr addrspace(1) 0 to i64
%2 = inttoptr i64 %1 to ptr addrspace(4)
%3 = inttoptr i64 %1 to ptr addrspace(4)
```

Reply:
Yes. Have you seen it in real codes?

Reply:
> Have you seen it in real codes?

Does it matter? Answering no here doesn't rule out the possibility. We can agree they are unlikely, however. If you are not going to cover this pattern you should at least add a comment about the caveat and put in some alarm (like add `assert(PtrToInt->hasOneUse());`) catch the case.

Reply:
This peephole is for a specific pattern that comes from the CUDA/HIP frontend, not meant to add complexity for the sake of covering unseen cases.  But you are right, a comment and assert should be good.